### ISSUE TITLE
Preconditioning

### ISSUE THREAD

**Issue by alphaville** [alphaville/optimization-engine#107]
(Implemented preconditioning in Python)

Implement scaling of the cost function and the constraints...

![image](https://user-images.githubusercontent.com/125415/107107788-45de2180-682b-11eb-9d88-62e8ae83c33c.jpeg)

where, for example,

![image](https://user-images.githubusercontent.com/125415/107107778-37900580-682b-11eb-9bca-0b73120ccd73.jpeg)

and

![image](https://user-images.githubusercontent.com/125415/107107850-9bb2c980-682b-11eb-8eca-f5edbf88a957.jpeg)

[which is not very straightforward as we don't compute the Jacobian of `F1` directly]

The weights can be easily computed in Rust.

**Comment by alphaville** [comment]
@korken89 This is why this is not so straightforward: we need to compute scaling parameters `w_f`, `w_1` and `w_2` such that the FBE becomes

![image](https://user-images.githubusercontent.com/125415/107103953-9481c080-6817-11eb-8a6a-9902e48452c0.jpeg)

Recall that in Rust, we only have access to `F1`, `F2` and `psi`, but not `f` and that `psi` as well as its gradient auto-generated.

The problem is not how we can compute the scaling parameters `w_f`, `w_1` and `w_2`, but how to actually scale `psi` and its gradient.

However, here is a possible easy and, in my opinion, clean solution: Casadi will generate code for the function

![image](https://user-images.githubusercontent.com/125415/107104248-b29bf080-6818-11eb-9686-d197a0ffb110.jpeg)

the argument `p` is already there: it is the user-defined parameter; the additional argument `w` will be the vector of scaling parameters. In fact this may or may not be part of `p`.

In other words:

![image](https://user-images.githubusercontent.com/125415/107104374-4bcb0700-6819-11eb-950d-7aef4a8b85cd.jpeg)

My proposal would be to use the existing vector `p` to pass `w` because it will be less disruptive.

**Comment by korken89** [comment]
Makes sense, I don't remember the interface we use right now, but I understand there is a limit on the interface and we cannot add more parameters? Or is it just simply a simple way to include the new scanning parameters?

**Comment by alphaville** [comment]
That's an overview of the various interfaces:

![image](https://user-images.githubusercontent.com/125415/107126902-7ca74c80-68aa-11eb-8345-51d1e2a3fbca.png)

**Comment by korken89** [comment]
Ah okey, it's an issue for the augmented Lagrangian. Does the penalty method have the same issue?

**Comment by alphaville** [comment]
@Arun-R-M @smokinmirror To push this forward, let us break it down into small tasks. Every task should correspond to a PR. 

Regarding DevOps:

- create a branch called `dev` out of `master` 
- for every PR, create a branch out of `dev` and after review, merge it into `dev` (not master)
- include unit tests where possible in every PR

Here is a list of tasks:

- [ ] In Python, generate a C file (via CasADi) with the preconditioning parameters (the w's and ρ). Do not change the number of parameters of the solver at this stage
- [ ] In `templates/icasadi/`, in `interface.c`, implement a wrapper in C around the auto-generated files. There are several examples in the same file. You may have to do minor edits in `casadi_memory.h`. You must test this implementation.
- [ ] Move on to `templates/icasadi/icasadi_lib.rs`. This generates the file `icasadi/lib.rs`. There you need to make an interface to allow you to call the C code from Rust. 
- [ ] Modify `templates/optimizer.rs`. This generates the main file `src/lib.rs` in the auto-generated solver. Focus on function `solve`

**Comment by korken89** [comment]
🚀 

Happy to see this getting some focus :D

**Comment by korken89** [comment]
Also, ping me if you want some feedback :)

**Comment by alphaville** [comment]
> Also, ping me if you want some feedback :)

Thank you, Emil. This is almost done now (we're testing it). I'll send you a review request soon.
